### DISCUSSION TITLE
Creating a custom outline

### DISCUSSION THREAD

**Discussion by CasparNuel** [typst/typst#2681] [Q&A] [answered]
Is it currently possible to create a custom outline?

Take for example I have this code:

```typ
#let qanda(question, answer) = {
    text("Q:")
    label("qanda")   // Typst compiler complains about this
    emph(question)
    linebreak()
    text("A:")
    text(answer)
    }

= An example

#qanda("What is 1+1?")[1+1 Is 2]

#lorem(20)

#qanda("What is 1 + 3")[1 + 3 Is 4]

#outline(title: "Questions", target: label("qanda")) // How do I achieve this
```

I get the error `error: cannot join content with label`. This does not seem like the right way to go.

Is there a way to create an outline that collects all questions in an easy to read list, so one can jump to the question after reading?

**Comment by ghost** [comment, marked as answer]
The error goes away with the following code, is this what you want to achieve?

```typst
#let qanda(question, answer) = {
    [Q: <qanda>]
    emph(question)
    linebreak()
    text("A:")
    text(answer)
}
```

**Reply by CasparNuel** [reply]
This does indeed remove the compilation error but it still results in an empty outline (bottom line of my MWE). This is expected as I have multiple identical labels pointing to different things. 


I am probably approaching this issue from the wrong angle but I have no idea how to achieve this in typst at the moment.

In essence I want to achieve the functionality of LaTeX's `\addcontentsline{toc}{chapter}{Dodatak}` and make it clickable in a **custom** TOC (for example: an outline that only concerns questions in the document that can be consulted at the end of a document).

**EDIT:**
It appears I was too fast. Your suggestion does in fact still result in a compilation error: `error: cannot outline text`.

Further suggestion to achieve this custom outline are definitely welcome!

**Reply by ghost** [reply]
Something like this is probably what you want to achieve:

```typst
#let question(question, answer) = {
    hide(box(height: 0pt, figure(
        none,
        kind: "question",
        supplement: [Question],
        caption: question,
        //numbering: none,
        outlined: true,
    )))
    heading(
        question,
        level: 1,
        numbering: none,
        supplement: [Question],
        outlined: false,
        bookmarked: true,
    )
    answer
}


#question([What is the first question?], [None yet])
#question([What is the second question?], [None...])


#outline(
    title: "Questions",
    target: figure.where(kind: "question"),
)
```
![questions](https://github.com/typst/typst/assets/138529922/32ecfa62-6134-4055-ad4e-aa7aec4217c5)

Note that `outline` still works when put at the top, even when the questions come later.

When using only `heading` and no `figure` and then filtering by `heading.where(supplement: [Question])` also works, but then the Questions also appear in the main TOC, what is probably not wanted.

**Reply by CasparNuel** [reply]
This does achieve what I want but seeing that it requires hiding a (sizeless) figure element it feels more hacky than I'd like. Maybe I should make a feature request to enable a less hacky way of achieving this.

Perhaps something like (just a quick thought, I'll think it through more if I end up creating a real feature request).

```typ
outlineable(
    kind,
    content . . . 
)
```
or perhaps a new parameter for outlineable content (text, figures, headings, ...).

Thank you anyways, this will do for now even though I'm not really a fan of this work-around :smile:

**Reply by PgBiel** [reply]
> Perhaps something like (just a quick thought, I'll think it through more if I end up creating a real feature request).
> 
> ```typst
> outlineable(
>     kind,
>     content . . . 
> )
> ```

I believe this will be made possible through #147 :)

(Although it will take a while to land still...!)

**Reply by CasparNuel** [reply]
It definitely could be a way to achieve this. Although I think a more general way of handling interaction with the current outline feature would be very welcome too.

I think it is a bit limited to only be able to include headings, tables and figures in the outline currently, as there are many other things that are also worth outlining (e.g. code listings)

**Reply by PgBiel** [reply]
> It definitely could be a way to achieve this. Although I think a more general way of handling interaction with the current outline feature would be very welcome too.

The idea is that outline targets a set of elements, so with custom elements you'd be able to have the outline target whatever you want :wink: 

However, you can already simulate this with figures instead of using headings, as the accepted solution shows. Just to give another example, you could do something like this:

```typ
#outline(target: figure.where(kind: "customthing"))

#let customthing(x) = figure(
  x,
  kind: "customthing",
  supplement: [Customized],
  caption: "A custom thing"
)

#customthing[Hello world]

#customthing(rect(width: 100pt, height: 100pt))
```

This produces:

![image with two custom figures showing up in the outline](https://github.com/typst/typst/assets/9021226/ccf611cf-f99c-4ca8-962a-d5d91432093f)

Maybe you could even omit the caption entirely there, if desired.

So, it's already possible to customize it... a bit... haha

But for entirely custom things you'd certainly need custom elements.

> I think it is a bit limited to only be able to include headings, tables and figures in the outline currently

For sure. Though figures are pretty versatile already - they aren't just for images! The idea behind figures is to be able to wrap anything in a way that can be referenced to and outlined. So this is more or less their purpose.

> as there are many other things that are also worth outlining (e.g. code listings)

Place them in figures. It's the recommended way to do that:

````typ
#outline(target: figure.where(kind: raw))

#figure(```js hello world```, caption: [A])

#figure(```js bye world```, caption: [B])
````

This will produce:

![image with code listings showing up in the outline](https://github.com/typst/typst/assets/9021226/1c1de3bf-c71d-4a66-b6d1-0bfde43e4c76)

**Reply by ghost** [reply]
The `hide` is not required if and only if all Typst codegen is super perfect. I assumed it is at least a bit buggy, so I added the `hide` to be on the safe side. The `hide` also tells the Typst compiler together with the 0pt height, that it does not need to write this content in the actual output. It probably does, though.

**Reply by CasparNuel** [reply]
> > It definitely could be a way to achieve this. Although I think a more general way of handling interaction with the current outline feature would be very welcome too.
> 
> The idea is that outline targets a set of elements, so with custom elements you'd be able to have the outline target whatever you want 😉
> 
> However, you can already simulate this with figures instead of using headings, as the accepted solution shows. Just to give another example, you could do something like this:
> 
> ```typst
> #outline(target: figure.where(kind: "customthing"))
> 
> #let customthing(x) = figure(
>   x,
>   kind: "customthing",
>   supplement: [Customized],
>   caption: "A custom thing"
> )
> 
> #customthing[Hello world]
> 
> #customthing(rect(width: 100pt, height: 100pt))
> ```
> 
> This produces:
> 
> ![image with two custom figures showing up in the outline](https://user-images.githubusercontent.com/9021226/284311085-ccf611cf-f99c-4ca8-962a-d5d91432093f.png)
> 
> Maybe you could even omit the caption entirely there, if desired.
> 
> So, it's already possible to customize it... a bit... haha
> 
> But for entirely custom things you'd certainly need custom elements.
> 
> > I think it is a bit limited to only be able to include headings, tables and figures in the outline currently
> 
> For sure. Though figures are pretty versatile already - they aren't just for images! The idea behind figures is to be able to wrap anything in a way that can be referenced to and outlined. So this is more or less their purpose.
> 
> > as there are many other things that are also worth outlining (e.g. code listings)
> 
> Place them in figures. It's the recommended way to do that:
> 
> ```typst
> #outline(target: figure.where(kind: raw))
> 
> #figure(```js hello world```, caption: [A])
> 
> #figure(```js bye world```, caption: [B])
> ```
> 
> This will produce:
> 
> ![image with code listings showing up in the outline](https://user-images.githubusercontent.com/9021226/284311670-1c1de3bf-c71d-4a66-b6d1-0bfde43e4c76.png)

This would still require changes to the current working of the outline, correct? As currently, there is not a way to make the outline point to a certain element, only to a "kind", or using a manual function. Is there currently a way to, say, outline all `emph`?

**Reply by PgBiel** [reply]
> This would still require changes to the current working of the outline, correct? As currently, there is not a way to make the outline point to a certain element, only to a "kind", or using a manual function. Is there currently a way to, say, outline all emph?

The problem is that elements have to be outlinable. What would an `emph` display on the outline? Sounds like that would vary a lot depending on your use case.  Hence why you need to wrap elements in figures, as figures let you specify a supplement, perhaps a caption, and provide numbering, reference and outline support automatically.

**Reply by CasparNuel** [reply]
I partially agree. I would assume an outline of anything contains a list of the "representation" of that element, with a reference to its location. An outline of just "image" elements would for example be a list of the literal images, clickable to jump to their location in the document.

Some element types would have specific representations (e.g., a figure's outline representation is its supplement + caption) or maybe defineable representations (`this => {this.content}`)

Doing this for every element is probably not feasible as there is a wide range of possible parameters and scaling (when e.g. outlining tables or images or visual lines) would be awkward at best, but more options than the current method would be welcome (e.g. outlining text elements is something that should be allowed imo without wrapping into figure, it is more an outline versatility thing than a figure thing for me)
